### PR TITLE
Revise register estimation algorithm used in inner normalization

### PR DESCRIPTION
**Pull request by liqiangxl**
For a detailed document see [this link](https://docs.google.com/document/d/1hZsLg-N11IKXq5vrqJAo-ot99t8daxLvHJN71XtKeBE/edit?usp=sharing).

**Issue:** An inaccurate estimation of register usage within the inner normalization scheduler results in low occupancy of 25%, and low bandwidth of 30% SOL, when fused with dropout.

**Fix:** This PR calculates the register count based on a target occupancy constrained with a minimum number to prevent significant register spills.
The corresponding code block is moved to a standalone function `estimateRegPerThread`:
(1) It takes two free parameters: 

- `target_occupancy`: minimum occupancy required to hide memory access latency.
- `register_overhead`: all registers used in a thread except those for persistent buffers.

(2) The function works as follows:
- Estimate register usage from target occupancy
- If it is not less than the minimum required cout (`overhead+buffer size`), use the estimated value
- Otherwise, estimate occupancy from the minimum required count and recalculate register available for each thread from the estimated occupancy. 

Note that, the register count each thread can use is always calculated from occupancy. The register estimation using `overhead+buffer size` just serves as a constrain on the occupancy the kernel can achieve.

**Results:** 
After the fix, the occupancies of many cases increased to above 37.5% and bandwidth increased by 1.2x to 1.4x. 

**Followed works:** 
Further performance boost will be done in another PR by optimizing the persistent batch size.

image

### CODE REVIEW

File: csrc/scheduler/normalization_inner.cpp [hunk 1]
```diff
@@ -177,6 +177,41 @@ void InnerPersistentKernelScheduler::computeHeuristics(
 
 namespace {
 
+// Calculate register usage based on target occupancy
```
[the remark this answers is not in the document]
Comment (naoyam) [reply csrc/scheduler/normalization_inner.cpp:180]: Please also mention that there are two major constraints we consider:

-We always guarantee the maxregistercount is at least as large as the buffer+overhead estimate
- We meet the desired occupancy but don't try to maximize it further. As long as it's as large as a given target occupancy, we consider it's good enough. The idea is that as long as we have a good enough occupancy, we should be able to saturate the memory bandwidth.

Within these constraints, we try to maximize the number of registers each thread can use.

[the remark this answers is not in the document]
Comment (liqiangxl) [reply csrc/scheduler/normalization_inner.cpp]: Thanks for the comments. revised accordingly.

**Comment by naoyam** [comment]
Have you run the CI tests?

**Review by naoyam — APPROVED** [review]
LGTM

**Comment by liqiangxl** [comment]
!build

**Comment by liqiangxl** [comment]
> Have you run the CI tests?

just did CI tests, all passed.

**Pull request merged**